src/lib.rs

```diff
@@ -160,60 +118,102 @@ pub fn get_jvm_dyn_lib_file_name() -> &'static str {
 ///
 /// If `JAVA_HOME` is not defined, the function tries to locate it using the `java` executable.
-pub fn locate_java_home() -> errors::Result<String> {
+pub fn locate_java_home() -> Result<String> {
     match &env::var("JAVA_HOME") {
         Ok(s) if s.is_empty() => do_locate_java_home(),
         Ok(java_home_env_var) => Ok(java_home_env_var.clone()),
         Err(_) => do_locate_java_home(),
     }
 }
 
-fn do_locate_java_home() -> errors::Result<String> {
-    // Prepare the command depending on the host
-    let command_str = if is_windows() {
-        "where"
-    } else if is_macos() {
-        "/usr/libexec/java_home"
-    } else {
-        "which"
-    };
+#[cfg(target_os = "windows")]
+fn do_locate_java_home() -> Result<String> {
+    let output = Command::new("where")
+        .arg("java")
+        .output()
+        .map_err(|e| JavaLocatorError::new(format!("Failed to run command `where` ({e})")))?;
+
+    let java_exec_path = std::str::from_utf8(&output.stdout)?
+        // Windows will return multiple lines if there are multiple `java` in the PATH.
+        .lines()
+        // The first line is the one that would be run, so take just that line.
+        .next()
+        .unwrap()
+        .trim();
 
-    let mut command = Command::new(command_str);
+    if java_exec_path.is_empty() {
+        return Err(JavaLocatorError::new(
+            "Java is not installed or not in the system PATH".into(),
+        ));
+    }
+
+    let mut home_path = follow_symlinks(java_exec_path);
+
+    // Here we should have found ourselves in a directory like /usr/lib/jvm/java-8-oracle/jre/bin/java
+    home_path.pop();
+    home_path.pop();
+
+    home_path
+        .into_os_string()
+        .into_string()
+        .map_err(|path| JavaLocatorError::new(format!("Java path {path:?} is invalid utf8")))
+}
 
-    if !is_macos() {
-        command.arg("java");
+#[cfg(target_os = "macos")]
+fn do_locate_java_home() -> Result<String> {
+    let output = Command::new("/usr/libexec/java_home")
+        .output()
+        .map_err(|e| {
+            JavaLocatorError::new(format!(
+                "Failed to run command `/usr/libexec/java_home` ({e})"
+            ))
+        })?;
+
+    let java_exec_path = std::str::from_utf8(&output.stdout)?.trim();
+
+    if java_exec_path.is_empty() {
+        return Err(JavaLocatorError::new(
+            "Java is not installed or not in the system PATH".into(),
+        ));
     }
 
-    let output = command.output().map_err(|error| {
-        let message = format!(
-            "Command '{}' is not found in the system PATH ({})",
-            command_str, error
-        );
-        errors::JavaLocatorError::new(&message)
-    })?;
-    let java_exec_path = String::from_utf8(output.stdout).map(|jp| {
-        let mut lines: Vec<&str> = jp.lines().collect();
-        if lines.len() > 1 {
-            println!(
-                "WARNING: java_locator found {} possible java locations: {}. Using the last one.",
```

Comment:
Please leave the warning here. The users need to be aware there are many versions found and maybe this is not what they want. They could just set the `JAVA_HOME` and make their choice.

Reply:
A library should not be dumping logs to stdout.
I've reintroduced the warning as writing to stderr which is not ideal but still better than stdout.
Ideally we would write it with something like the log crate https://docs.rs/log/latest/log/
But I'm not going to block the PR on this.

Reply:
Hmmm noticed some issues here, looking into it.

Reply:
Fixed the issues.
The warning only applies to windows as the unix `which` command will only return a single path regardless of how many executables there are in PATH with the requested name.